### PR TITLE
Allow session properties defaults in transaction

### LINKED ISSUE #6642: Transactions does not work when session properties are set using Session Property Manager
When session properties are configured in presto server, transactions does not work and throws the issue.
Session properties cannot be overridden once a transaction is active
	at com.google.common.base.Preconditions.checkState(Preconditions.java:507)
	at io.prestosql.Session.withDefaultProperties(Session.java:381)
	at io.prestosql.server.SessionPropertyDefaults.newSessionWithDefaultProperties(SessionPropertyDefaults.java:109)
	at io.prestosql.execution.SqlQueryManager.createQueryInternal(SqlQueryManager.java:349)
	at io.prestosql.execution.SqlQueryManager.lambda$createQuery$4(SqlQueryManager.java:301)
	at io.prestosql.$gen.Presto_305____20201130_094529_1.run(Unknown Source)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)

https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/Session.java#L397-L400

### PR DESCRIPTION
**Pull request by aczajkowski**
Resolves #6642

### CODE REVIEW

**Review by electrum — CHANGES_REQUESTED** [review]
This PR seems to do the same as #9025, with changes requested by by @findepi to add tests, and the author requesting help on how to write a test. Please work with @findinpath to write the appropriate tests requested in https://github.com/trinodb/trino/pull/9025/files#r698286027

**Comment by aczajkowski** [comment]
> This PR seems to do the same as #9025, with changes requested by by @findepi to add tests, and the author requesting help on how to write a test. Please work with @findinpath to write the appropriate tests requested in https://github.com/trinodb/trino/pull/9025/files#r698286027

@electrum I didn't know about the PR. I've compared them now. And yes the solution is similar as it was described how to fix in comment over the check we want to remove. So i would expect it to be the similar in every case
https://github.com/trinodb/trino/blob/ef4b88729d572c25ddeccae420b082cadcfbcf63/core/trino-main/src/main/java/io/trino/Session.java#L403

In my PR i've made more changes:
- removal of redundant code in this case by extracting validateSystemProperties and validateCatalogProperties
- removal of redundant argument in SessionPropertyManager. validateCatalogSessionProperty
- Integration test adding test code coverage

Please advice how to approach it.

File: core/trino-main/src/main/java/io/trino/Session.java [hunk 1]
```diff
@@ -395,20 +382,17 @@ public Session beginTransactionId(TransactionId transactionId, TransactionManage
                 Optional.of(isAutoCommitContext));
     }
 
-    public Session withDefaultProperties(Map<String, String> systemPropertyDefaults, Map<String, Map<String, String>> catalogPropertyDefaults)
+    public Session withDefaultProperties(AccessControl accessControl, Map<String, String> systemPropertyDefaults, Map<String, Map<String, String>> catalogPropertyDefaults)
```
Comment (findinpath) [inline core/trino-main/src/main/java/io/trino/Session.java:385]: minor: Why do you place `accessControl` parameter on the 1st position?
The method's name is `withDefaultProperties`, so I'd place `accessControl` actually (from precedence) on the last parameter position

File: plugin/trino-session-property-managers/src/test/java/io/trino/plugin/session/TestSessionPropertyManagerInTransaction.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        // Assure session property defaults are applied
+        assertQuery(
+                "SHOW SESSION LIKE '" + QUERY_MAX_EXECUTION_TIME + "'",
+                "VALUES('" + QUERY_MAX_EXECUTION_TIME + "','8h', '100.00d', 'varchar', 'Maximum execution time of a query')");
+        // Perform operation in transaction
+        newTransaction()
+                .execute(getSession(), session -> {
+                    getQueryRunner().execute(session, "CREATE SCHEMA test");
+                });
+        // Assure operation was successful
```
Comment (findinpath) [inline plugin/trino-session-property-managers/src/test/java/io/trino/plugin/session/TestSessionPropertyManagerInTransaction.java:76]: Assure operation was successful -> Ensure that the previous statement was successful

File: plugin/trino-session-property-managers/src/test/java/io/trino/plugin/session/TestSessionPropertyManagerInTransaction.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                        // This is required when using MinIO which requires path style access
+                        .put("hive.s3.path-style-access", "true")
+                        .put("hive.insert-existing-partitions-behavior", "OVERWRITE")
+                        .put("hive.non-managed-table-writes-enabled", "true")
+                        .build());
+        queryRunner.installPlugin(new SessionPropertyConfigurationManagerPlugin());
+        queryRunner.getSessionPropertyDefaults().loadConfigurationManager(CONFIG_FILE.getAbsoluteFile());
+        return queryRunner;
+    }
+
```
Comment (findinpath) [inline plugin/trino-session-property-managers/src/test/java/io/trino/plugin/session/TestSessionPropertyManagerInTransaction.java:63]: Pls do take into account the suggestion from @findepi : https://github.com/trinodb/trino/pull/9025/files#r698286027

**Comment by electrum** [comment]
Sorry, let me clarify my comment. I didn't meant to imply you knew about the other one or copied it -- quite the opposite, it's certainly frustrating to do duplicate work.

We generally prefer to merge the earlier PR, unless one approach is substantially better, in order to be fair and give credit to the earlier author. If you favor the latter one, it can appear that you're taking credit for their work, and it's just generally not a good experience.

It would be helpful if you could work with @findinpath on how to write the tests that @findepi asked for, since you have "access" to people who are knowledgeable about that. @findinpath it's up to you on how you want to proceed. The removal of the redundant code is good, so we could integrate that into the original PR, or do it as a follow up.

**Comment by aczajkowski** [comment]
@electrum Let me alighn with @findinpath . We will find a path forward. 
Thank you for your help.

File: core/trino-main/src/main/java/io/trino/Session.java (hunk 1 above)
Comment (aczajkowski) [reply core/trino-main/src/main/java/io/trino/Session.java]: Applied

File: core/trino-main/src/main/java/io/trino/testing/AllowAllAccessControlManager.java [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+public class AllowAllAccessControlManager
+        implements AccessControl
+{
+    @Override
+    public void checkCanSetUser(Optional<Principal> principal, String userName)
+    {}
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/testing/AllowAllAccessControlManager.java:36]: nit: can you put `{}` in the same lie as method name?

File: core/trino-main/src/test/java/io/trino/server/TestSessionPropertyDefaults.java [hunk 5]
```diff
@@ -39,28 +44,34 @@
     @Test
     public void testApplyDefaultProperties()
     {
-        SessionPropertyDefaults sessionPropertyDefaults = new SessionPropertyDefaults(TEST_NODE_INFO);
+        SessionPropertyDefaults sessionPropertyDefaults = new SessionPropertyDefaults(TEST_NODE_INFO, new AllowAllAccessControlManager());
```
Comment (losipiuk) [inline core/trino-main/src/test/java/io/trino/server/TestSessionPropertyDefaults.java:47]: Most chagnes here seem unrelated to `Inject AccessControl to SessionPropertyDefaults`

File: core/trino-main/src/main/java/io/trino/metadata/SessionPropertyManager.java [hunk 6]
```diff
@@ -209,9 +209,9 @@ public void validateSystemSessionProperty(String propertyName, String propertyVa
         decodePropertyValue(propertyName, propertyValue, propertyMetadata.getJavaType(), propertyMetadata);
     }
 
-    public void validateCatalogSessionProperty(CatalogName catalog, String catalogName, String propertyName, String propertyValue)
+    public void validateCatalogSessionProperty(CatalogName catalog, String propertyName, String propertyValue)
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/metadata/SessionPropertyManager.java:212]: there were no call spots?

Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/metadata/SessionPropertyManager.java:212]: Also I think it may have a purpose (or maybe just used to have).
@findepi do you know maybe why we often use `String` and `CatalogName` aside together?

File: core/trino-main/src/main/java/io/trino/server/SessionPropertyDefaults.java [hunk 7]
```diff
@@ -71,11 +71,15 @@ public void addConfigurationManagerFactory(SessionPropertyConfigurationManagerFa
     public void loadConfigurationManager()
             throws IOException
     {
-        File configFile = CONFIG_FILE.getAbsoluteFile();
+        loadConfigurationManager(CONFIG_FILE.getAbsoluteFile());
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/server/SessionPropertyDefaults.java:74]: `Add test for default session properties in transaction` should be squashed with commit which adds new functinality.

File: plugin/trino-session-property-managers/src/test/java/io/trino/plugin/session/TestSessionPropertyManagerInTransaction.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+public class TestSessionPropertyManagerInTransaction
+        extends AbstractTestQueryFramework
+{
+    public static final File CONFIG_FILE = new File("src/test/resources/io/trino/plugin/session/file/session-property-config.properties");
+
+    private HiveMinioDataLake dockerizedS3DataLake;
+
+    @Override
+    protected QueryRunner createQueryRunner()
```
Comment (losipiuk) [inline plugin/trino-session-property-managers/src/test/java/io/trino/plugin/session/TestSessionPropertyManagerInTransaction.java:39]: can we use sth more lightweight than Hive here?

**Review by losipiuk — COMMENTED** [review]
Seems legit but commit boundaries are off.

File: core/trino-main/src/test/java/io/trino/server/TestSessionPropertyDefaults.java (hunk 5 above)
Comment (aczajkowski) [reply core/trino-main/src/test/java/io/trino/server/TestSessionPropertyDefaults.java:47]: Right, my bad. It should be in "Allow session properties override once in transaction"

Comment (aczajkowski) [reply core/trino-main/src/test/java/io/trino/server/TestSessionPropertyDefaults.java:47]: Fixed

File: core/trino-main/src/main/java/io/trino/metadata/SessionPropertyManager.java (hunk 6 above)
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/metadata/SessionPropertyManager.java:212]: i do not. maybe just a remnant from before https://github.com/trinodb/trino/pull/406? cc @kokosing

File: core/trino-main/src/main/java/io/trino/server/SessionPropertyDefaults.java (hunk 7 above)
Comment (aczajkowski) [reply core/trino-main/src/main/java/io/trino/server/SessionPropertyDefaults.java:74]: Yes. It was like this originally, but i wanted to split it to give author to @findinpath. How it's now.

File: plugin/trino-session-property-managers/src/test/java/io/trino/plugin/session/TestSessionPropertyManagerInTransaction.java (hunk 8 above)
Comment (aczajkowski) [reply plugin/trino-session-property-managers/src/test/java/io/trino/plugin/session/TestSessionPropertyManagerInTransaction.java:39]: I was not able to find anything that support transactions.
In memory, balckhole and mysql (which i tired) transactions are not allowed. Then I've check in code that only Hive have this unblocked.
Byt the way there is not information about that in documentation. I mean which connectors support and which not.

File: core/trino-main/src/main/java/io/trino/metadata/SessionPropertyManager.java (hunk 6 above)
Comment (aczajkowski) [reply core/trino-main/src/main/java/io/trino/metadata/SessionPropertyManager.java:212]: I haven't found more locations, but yes I would like to here @findepi opinion. It's easy to drop this commit.

Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/metadata/SessionPropertyManager.java:212]: > I haven't found more locations

I mean this commit changes method signature but does not change any call-spots. This seems wrong (unless method is not needed at all). I think you are actually changing call-spots in next commit.

Comment (aczajkowski) [reply core/trino-main/src/main/java/io/trino/metadata/SessionPropertyManager.java:212]: 🤦‍♂️ yes my bad. IT seems my yesterday rebasing made some mess.
Fixed now.

**Comment by aczajkowski** [comment]
@electrum as suggested by @findinpath https://github.com/trinodb/trino/pull/9025#issuecomment-960190559 We will merge my PR. I authored commit with actual fix for @findinpath .
Please let me know if this is ok and eventually remove request-changes.

File: plugin/trino-session-property-managers/src/test/java/io/trino/plugin/session/TestSessionPropertyManagerInTransaction.java (hunk 3 above)
Comment (aczajkowski) [reply plugin/trino-session-property-managers/src/test/java/io/trino/plugin/session/TestSessionPropertyManagerInTransaction.java:63]: Discussed with @findepi and for now we will leave it without this test.

File: core/trino-main/src/test/java/io/trino/server/TestSessionPropertyDefaults.java [hunk 9]
```diff
@@ -41,27 +45,33 @@
     public void testApplyDefaultProperties()
     {
         SessionPropertyDefaults sessionPropertyDefaults = new SessionPropertyDefaults(TEST_NODE_INFO, new AllowAllAccessControlManager());
+
+        SessionPropertyManager sessionPropertyManager = new SessionPropertyManager();
```
Comment (losipiuk) [inline core/trino-main/src/test/java/io/trino/server/TestSessionPropertyDefaults.java:49]: nit: test cleanups seem unrelated. Maybe be extracted as separate commit.

**Review by losipiuk — COMMENTED** [review]
LGMT though I am still not 100% sure about `Remove redundant argument in SessionPropertyManager.validateCatalogSessionProperty`.

Can you double check that removed parameter always matches the `catalog.getCatalogName` ?

File: core/trino-main/src/test/java/io/trino/server/TestSessionPropertyDefaults.java (hunk 9 above)
Comment (aczajkowski) [reply core/trino-main/src/test/java/io/trino/server/TestSessionPropertyDefaults.java:49]: Unfortunately those changes are required here. We need to use real properties from now on. As we started validating them when setting defaults. And as now they are not self explanatory strings I needed to add comments.

**Comment by aczajkowski** [comment]
> LGMT though I am still not 100% sure about `Remove redundant argument in SessionPropertyManager.validateCatalogSessionProperty`.
> 
> Can you double check that removed parameter always matches the `catalog.getCatalogName` ?

Oryginaly it was used in single location in Session.
When looking at code https://github.com/trinodb/trino/blob/aeb753aed2f04d6a6d520edb5645f03bb24e1721/core/trino-main/src/main/java/io/trino/Session.java#L330
CatalogName is returned from TransactionManger basing on catalogName.
And in transaction manager it's created/stored like so.
https://github.com/trinodb/trino/blob/aeb753aed2f04d6a6d520edb5645f03bb24e1721/core/trino-main/src/main/java/io/trino/transaction/InMemoryTransactionManager.java#L431
So this seems to be always aligned.

**Review by losipiuk — APPROVED** [review]

File: plugin/trino-session-property-managers/src/test/java/io/trino/plugin/session/TestSessionPropertyManagerInTransaction.java (hunk 8 above)
Comment (findinpath) [reply plugin/trino-session-property-managers/src/test/java/io/trino/plugin/session/TestSessionPropertyManagerInTransaction.java:39]: I also had a hard time finding out in the documentation which connectors support transactions.

**Comment by losipiuk** [comment]
@electrum good to go? Your concern regarding the other PR is resolved.

File: plugin/trino-session-property-managers/src/test/java/io/trino/plugin/session/TestSessionPropertyManagerInTransaction.java (hunk 8 above)
Comment (mosabua) [reply plugin/trino-session-property-managers/src/test/java/io/trino/plugin/session/TestSessionPropertyManagerInTransaction.java:39]: From all I know no connector supports transaction and that is why it is not documented in SQL support section for the connectors ... at least thats what I remember from talking to @electrum about it

Comment (aczajkowski) [reply plugin/trino-session-property-managers/src/test/java/io/trino/plugin/session/TestSessionPropertyManagerInTransaction.java:39]: @mosabua  From my finding and all the discussion around this PR Hive Connector does.

**Comment by losipiuk** [comment]
@electrum merging this one as your only concern about other PR existing was resolved, and authorship of @findinpath is resolved on main commit which adds feature. Thanks @aczajkowski and @findinpath for resolving that.

File: plugin/trino-session-property-managers/src/test/java/io/trino/plugin/session/TestSessionPropertyManagerInTransaction.java (hunk 8 above)
Comment (mosabua) [reply plugin/trino-session-property-managers/src/test/java/io/trino/plugin/session/TestSessionPropertyManagerInTransaction.java:39]: Okay.. necessary follow up then - https://github.com/trinodb/trino/pull/10050 .. please review
